bson/src/main/org/bson/codecs/pojo/DiscriminatorLookup.java

```diff
@@ -55,30 +66,92 @@
+            if (Modifier.isInterface(classModel.getType().getModifiers()) || Modifier.isAbstract(classModel.getType().getModifiers())) {
+                String packageName = classModel.getType().getPackage() != null ? classModel.getType().getPackage().getName() : "";
+                if (!packageName.isEmpty()) {
+                    searchPackageForBsonDiscriminator(packageName);
+                }
+            }
         }
     }
 
-    private Class<?> getClassForName(final String discriminator) {
+    private Class<?> searchPackages(final String discriminator) {
+        Class<?> clazz = null;
+        for (String packageName : packages) {
+            clazz = getClassForName(packageName + "." + discriminator);
+            if (clazz != null) {
+                return clazz;
+            }
+        }
+        for (String packageName : packages) {
+            searchPackageForBsonDiscriminator(packageName);
+            if (discriminatorClassMap.containsKey(discriminator)) {
+                return discriminatorClassMap.get(discriminator);
+            }
+        }
+        return clazz;
+    }
+
+    private void searchPackageForBsonDiscriminator(final String packageName) {
+        getClassesInPackageWithBsonDiscriminator(packageName)
+                .forEach(c -> discriminatorClassMap.put(c.getAnnotation(BsonDiscriminator.class).value(), c));
+    }
+
+    private static Class<?> getClassForName(final String className) {
         Class<?> clazz = null;
         try {
-            clazz = Class.forName(discriminator);
+            clazz = Class.forName(className);
         } catch (final ClassNotFoundException e) {
             // Ignore
         }
         return clazz;
     }
 
-    private Class<?> searchPackages(final String discriminator) {
-        Class<?> clazz = null;
-        for (String packageName : packages) {
-            clazz = getClassForName(packageName + "." + discriminator);
-            if (clazz != null) {
-                return clazz;
+    private static List<Class<?>> getClassesInPackageWithBsonDiscriminator(final String packageName) {
+        if (!PACKAGE_CLASS_MAP.containsKey(packageName)) {
+            PACKAGE_CLASS_MAP.put(packageName,
+                    getClassNamesInPackage(packageName).stream()
+                    .map(DiscriminatorLookup::getClassForName)
+                    .filter(c -> c != null && c.isAnnotationPresent(BsonDiscriminator.class))
+                    .collect(Collectors.toList()));
+        }
+        return PACKAGE_CLASS_MAP.get(packageName);
+    }
+
+    private static List<String> getClassNamesInPackage(final String packageName) {
+        String packagePath = packageName.replaceAll("\\.", File.separator);
+        List<String> classNames = new ArrayList<>();
+        String[] classPathEntries = System.getProperty("java.class.path", "").split(System.getProperty("path.separator", File.separator));
```

Comment:
This reminds me of all the problems we had just trying to find the driver version in the driver's own MANIFEST.MF, before finally giving up and generating a class that contained the driver version.  

I don't remember all the details, but I think some of it had to do with the fact that some application servers use custom ClassLoader implementation that perhaps don't even use the class path.  And then there is OSGi.  And the Java 9+ module path.

Is there another way to accomplish the goal besides scanning the class path?